keps/sig-api-machinery/4793-revise-apf-default-configuration/README.md

```diff
@@ -0,0 +1,992 @@
+-->
+
+Many features have been developed and use cases surfaced since the introduction
+of [KEP-1040]. These prompt us to revise APF default configuration to better
+make use of APF for better performance and reliability of API Server.
+
+For example, we have learned some priority levels are more critical than other
+priority levels as when critical requests fail, the whole Kubernetes control
+plane may become unavailable (e.g. node-controller become unavailable and no new
+Nodes can be registered due to node-controller's leader election requests being
+failed). For such priority levels, we propose to increase their nominal shares
+to avoid failing critical requests and lendable percentage as well to avoid
+wasting shares.
+
+We also have learned that Event requests are plentiful in upscaling but they are
+resilient to unavailability of API Server. So we can deprioritize them when API
+Server is under stress.
+
+[KEP-1040]: https://github.com/kubernetes/enhancements/blob/97713189b3107b41c4c19505d04aa7ef22df063b/keps/sig-api-machinery/1040-priority-and-fairness/README.md
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+-   Better overload protection in perspective of whole Kubernetes control plane
+    instead of just API Server alone.
+-   More performant in cluster upscaling (e.g. increasing node count, deploying
+    pods)
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+-   There will be no fundamental changes in APF's own implementation
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The APF feature provides flow control to API Server by introducing a limit on
+how much a priority level can consume the concurrency. In addition to that,
+borrowing is also instroduced to better utilize spare concurrency. The limits on
+borrowing are two-sided: a given priority level has a limit on how much it may
+borrow and a limit on how much may be borrowed from it. Worth mentioning, APF
+controller periodically populates current concurrency shares for each priority
+level (i.e. concurrency share that can be borrowed or need to be borrowed) by
+taking past usage into consideration. So in combination of those two aspects, a
+priority level's Nomial Share is only an initial value and the actual share
+might increase or decrease according to actual usage with borrowing. However,
+the owner priority level should have full Nominal Share as its current
+concurrency share if in full utilization.
+
```

Comment:
My main question that is not touched on in this KEP is: how are we going to test it/validate that we will not visibly break someone.

I can easily buy that there are cases where this configuration helps a lot.
But I don't know how we mitigate a negative impact on some other setups.

Reply:
Good callout and this is very difficult as this KEP increases the SUM of nominal shares so it will definite dilute any custom APF configuration therefore visible user impacts.

I'm thinking out loud -- we can be only certain that it won't break k8s-promised scalability and performance characteristic (e.g. https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md) since there are tests to verify. Then we will be precautious on graduating this KEP to have user feedback to evaluate the risk of breaking existing configuration and provide mitigations. Though I admit that this is not ideal.

Reply:
> since there are tests to verify. 

Tests are never perfect.

> Then we will be precautious on graduating this KEP to have user feedback to evaluate the risk of breaking existing configuration and provide mitigations. Though I admit that this is not ideal.

Up until this is enabled by default, noone uses it. If we enable it by default, it's too late, because we already break someone.
This is not a valid mitigation strategy....